# PDS: Endterm - Practical part

### Date: February 26

### Time: 15:00 - 16:30


**Instructions:** 
* Write your full name and group number in the title of file;
* All answers should be filled in the notebook and then submitted to Moodle;
* Try to use fewer cells for compilation (for example, you can use print for several answers, instead of printing each answer on a separate cell)
* **IMPORTANT NOTE:** you will get points only for correct answers, not for trials or wrong answers. Your notebook should run if I will press **"Restart Kernel"** and then press **"Run All"**. I am going to check answers only after doing that. If the code do not compile, you will get zero. 

In [1]:
import pandas as pd
import numpy as np

## Problem 1 (1 point)

Create a function that takes non empty string as input, and returns Boolean output, which tells if a string is a polindrom. Polindrom is a string that has the same spelling from the left to the right and vice-versa. One symbol strings are also polyndroms. For example, 'aba' is a polindrom, while 'abc' is not a polindrom.

Use the following strings to check created function: 
1. 'a'
2. 'abcd'
3. 'abcdcba'
4. 'lol'
5. 'data'

In [2]:
def check_palindrome(s):
    return s == s[::-1]
 
s1 = 'a'
s2 = 'abcd'
s3 = 'abcdcba'
s4 = 'lol'
s5 = 'data'


print(check_palindrome(s1))
print(check_palindrome(s2))
print(check_palindrome(s3))
print(check_palindrome(s4))
print(check_palindrome(s5))

True
False
True
True
False


## Problem 2 (2 points)
Create a function that takes non empty string as input, and returns Boolean output, which tells if a string is an isogram. Isogram is a string that does not have repeating symbols. For example, 'abcd' is an isogram, while 'abca' is not an isogram.

Use the following strings to check created function: 
1. 'abcdcba'
2. 'abcdefg'
3. 'bowling'
4. 'lol'
5. 'data'

In [3]:
def check_isogram(str1):
    return len(str1) == len(set(str1.lower()))

s1 = 'abcdcba'
s2 = 'abcdefg'
s3 = 'bowling'
s4 = 'lol'
s5 = 'data'

print(check_isogram(s1))
print(check_isogram(s2))
print(check_isogram(s3))
print(check_isogram(s4))
print(check_isogram(s5))

False
True
True
False
False


## Problem 3 (2 points)
You have an array of numbers and one number (k). Create a function, that moves this number k to the end the array, and then returns changed array. Function should perform all changes inplace, i.e. movements should take place inside the array without creating intermediary array, ignoring the order of numbers other than k. For example, array = [3,2,3,3,4,3,5,1,3,2] and k = 3, then after applying function you should get [2,4,5,1,2,3,3,3,3,3].

Use the following inputs to check created function:
1. [1,1,2] and 1
2. [1,2,2,1,2,1] and 2
3. [3,2,3,3,4,3,5,1,3,2] and 3


In [4]:
def move_number_k(arr,k): 
    size = len(arr)
    count = 0 

    for i in range(size): 
        if arr[i] != k: 
            arr[count] = arr[i] 
            count+=1

    while count < size: 
        arr[count] = k
        count += 1
          
arr1 = [1,1,2] 
arr2 = [1,2,2,1,2,1]
arr3 = [3,2,3,3,4,3,5,1,3,2]


k1 = 1
k2 = 2
k3 = 3


move_number_k(arr1,k1) 
print(arr1) 
move_number_k(arr2,k2) 
print(arr2) 
move_number_k(arr3,k3) 
print(arr3) 

[2, 1, 1]
[1, 1, 1, 2, 2, 2]
[2, 4, 5, 1, 2, 3, 3, 3, 3, 3]


## Problem 4 (5 points)
For Coursera dataset (coursea_data.csv), do the tasks below.

Link to dataset: https://www.kaggle.com/siddharthm1698/coursera-course-dataset

### 1. Find three the most commonly appearing combination of certificate type and course difficulty. Show cerfiticate type, course difficulty and number of courses. (1 point)

In [5]:
df = pd.read_csv('coursera_data.csv')
df.head()

Unnamed: 0                                       course_title  \
0         134  (ISC)² Systems Security Certified Practitioner...   
1         743  A Crash Course in Causality:  Inferring Causal...   
2         874                     A Crash Course in Data Science   
3         413                            A Law Student's Toolkit   
4         635                A Life of Happiness and Fulfillment   

          course_organization course_Certificate_type  course_rating  \
0                      (ISC)²          SPECIALIZATION            4.7   
1  University of Pennsylvania                  COURSE            4.7   
2    Johns Hopkins University                  COURSE            4.5   
3             Yale University                  COURSE            4.7   
4   Indian School of Business                  COURSE            4.8   

  course_difficulty course_students_enrolled  
0          Beginner                     5.3k  
1      Intermediate                      17k  
2             Mixed                     130k  
3             Mixed                      91k  
4             Mixed                     320k

In [6]:
new = df.groupby(['course_Certificate_type','course_difficulty']).size().sort_values(ascending=False)
new.head(3)

course_Certificate_type  course_difficulty
COURSE                   Beginner             282
SPECIALIZATION           Beginner             196
COURSE                   Mixed                187
dtype: int64

### 2. Among 'Data Science' courses, find the course with the 2nd highest rating among Intermediate level courses. For obtained course, show its title, organization, and number of students enrolled. (1 point)

In [7]:
df['course_title'] = df['course_title'].str.lower()
new = df[df['course_title'].str.contains('data science')]
new = new[new['course_difficulty'] == 'Intermediate']
new = new.sort_values('course_rating', ascending = False)
new[1:2]

Unnamed: 0                   course_title course_organization  \
55         637  applied data science capstone                 IBM   

   course_Certificate_type  course_rating course_difficulty  \
55                  COURSE            4.7      Intermediate   

   course_students_enrolled  
55                      42k

### 3. For organizations from Ivy league (list provided below), find the one that has the highest presence in Coursera, i.e. the largest number of courses. For obtained organization, calculate average the rating for each of the certificate types. Show certificate type with mean rating.  (1 point)

In [8]:
ivy_list = ['Brown University', 'Columbia University', 'Cornell University', 'Dartmouth College', 'Harvard University', 'University of Pennsylvania', 'Princeton University', 'Yale University']

new = df[df['course_organization'].isin(ivy_list)]
org = new['course_organization'].max()
new = new[new['course_organization'] == org]

new = new.groupby('course_Certificate_type')['course_rating'].mean()
new

course_Certificate_type
COURSE            4.83125
SPECIALIZATION    4.60000
Name: course_rating, dtype: float64

### 4. Create a new column course_students_enrolled_new, which is the same as course_students_enrolled, but with numeric data type. To do that multiply the ones that ends to 'k' by 1000 and 'm' by 1000000. Show first 10 rows of the updated dataframe. (1 point)

*Hint: you might need to create a function, and then use apply() method.*

In [12]:
df['flag1'] = df['course_students_enrolled']
df['flag2'] = df['course_students_enrolled'].replace(r'[km]+$', '', regex=True).astype(float)

df.flag1 = df.course_students_enrolled.str.extract(r'[\d\.]+([km]+)', expand=False).fillna(1).replace(['k','m'], [10**3, 10**6]).astype(int)
df['course_students_enrolled_new'] = df['flag1'] * df['flag2']


df = df.drop(columns=['flag1', 'flag2'])

df.head(10)

Unnamed: 0                                       course_title  \
0         134  (isc)² systems security certified practitioner...   
1         743  a crash course in causality:  inferring causal...   
2         874                     a crash course in data science   
3         413                            a law student's toolkit   
4         635                a life of happiness and fulfillment   
5         661  adhd: everyday strategies for elementary students   
6          54                                    ai for everyone   
7         488                           ai for medical treatment   
8          58                        ai foundations for everyone   
9         129                           ai for medical diagnosis   

          course_organization course_Certificate_type  course_rating  \
0                      (ISC)²          SPECIALIZATION            4.7   
1  University of Pennsylvania                  COURSE            4.7   
2    Johns Hopkins University                  COURSE            4.5   
3             Yale University                  COURSE            4.7   
4   Indian School of Business                  COURSE            4.8   
5       University at Buffalo                  COURSE            4.7   
6             deeplearning.ai                  COURSE            4.8   
7             deeplearning.ai                  COURSE            4.8   
8                         IBM          SPECIALIZATION            4.7   
9             deeplearning.ai                  COURSE            4.7   

  course_difficulty course_students_enrolled  course_students_enrolled_new  
0          Beginner                     5.3k                        5300.0  
1      Intermediate                      17k                       17000.0  
2             Mixed                     130k                      130000.0  
3             Mixed                      91k                       91000.0  
4             Mixed                     320k                      320000.0  
5          Beginner                      39k                       39000.0  
6          Beginner                     350k                      350000.0  
7      Intermediate                     2.4k                        2400.0  
8          Beginner                      61k                       61000.0  
9      Intermediate                      12k                       12000.0

### 5. Assuming that all enrolled students will earn a certificate from the course, count how many certificates will be issued for each certificate type. Then order number of certificates issued in ascending order. And find share as a percentage (for example, 25.6) of each certificate type. Show result as a dataframe with columns certificate type, corresponding number and share. (1 point)

---

### Evaluation form (Keep it!)

| Question | Mark     | Comment   
|:-------:|:--------:|:----------------------
| 1       |   1/1    | Correct
| 2       |   2/2    | Correct
| 3       |   2/2    | Correct
| 4.1     |   1/1    | Correct
| 4.2     |   1/1    | Correct
| 4.3     |   1/1    | Correct
| 4.4     |   1/1    | Correct
| 4.5     |   1/1    | Correct
|**Total**|**10/10**  | 
